Exception when creating Google Cloud Storage BLOB.

Thanks for creating an issue! Please fill out this form so we can be
sure to have all the information we need, and to minimize back and forth.

Nexus: 3.13.0-01

* **What are you trying to do?**
First time usage, trying to add an already created Google Blob store.  Getting the following exception when I try to create it in nexus

```
2018-08-15 18:40:23,217+0000 ERROR [qtp1779474946-201] sherd org.sonatype.nexus.extdirect.internal.ExtDirectServlet - Failed to invoke action method: coreui_Blobstore.create, java-method: org.sonatype.nexus.coreui.BlobStoreComponent.create
org.sonatype.nexus.blobstore.api.BlobStoreException: BlobId: null, Unable to initialize blob store bucket: maven-storage
	at org.sonatype.nexus.blobstore.gcloud.internal.GoogleCloudBlobStore.init(GoogleCloudBlobStore.java:354)
	at org.sonatype.nexus.repository.internal.blobstore.BlobStoreManagerImpl.newBlobStore(BlobStoreManagerImpl.java:213)
	at org.sonatype.nexus.repository.internal.blobstore.BlobStoreManagerImpl.create(BlobStoreManagerImpl.java:156)
	at org.sonatype.nexus.common.stateguard.MethodInvocationAction.run(MethodInvocationAction.java:39)
	at org.sonatype.nexus.common.stateguard.StateGuard$GuardImpl.run(StateGuard.java:270)
	at org.sonatype.nexus.common.stateguard.GuardedInterceptor.invoke(GuardedInterceptor.java:53)
	at org.sonatype.nexus.blobstore.api.BlobStoreManager$create$0.call(Unknown Source)
	at org.sonatype.nexus.coreui.BlobStoreComponent.create(BlobStoreComponent.groovy:88)
	at com.palominolabs.metrics.guice.ExceptionMeteredInterceptor.invoke(ExceptionMeteredInterceptor.java:49)
	at com.palominolabs.metrics.guice.TimedInterceptor.invoke(TimedInterceptor.java:47)
	at org.sonatype.nexus.validation.internal.ValidationInterceptor.invoke(ValidationInterceptor.java:53)
	at org.apache.shiro.guice.aop.AopAllianceMethodInvocationAdapter.proceed(AopAllianceMethodInvocationAdapter.java:49)
	at org.apache.shiro.authz.aop.AuthorizingAnnotationMethodInterceptor.invoke(AuthorizingAnnotationMethodInterceptor.java:68)
	at org.apache.shiro.guice.aop.AopAllianceMethodInterceptorAdapter.invoke(AopAllianceMethodInterceptorAdapter.java:36)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.softwarementors.extjs.djn.router.dispatcher.DispatcherBase.invokeJavaMethod(DispatcherBase.java:142)
	at com.softwarementors.extjs.djn.router.dispatcher.DispatcherBase.invokeMethod(DispatcherBase.java:133)
	at org.sonatype.nexus.extdirect.internal.ExtDirectServlet$3.invokeMethod(ExtDirectServlet.java:236)
	at com.softwarementors.extjs.djn.router.dispatcher.DispatcherBase.dispatch(DispatcherBase.java:63)
	at com.softwarementors.extjs.djn.router.processor.standard.StandardRequestProcessorBase.dispatchStandardMethod(StandardRequestProcessorBase.java:73)
	at com.softwarementors.extjs.djn.router.processor.standard.json.JsonRequestProcessor.processIndividualRequest(JsonRequestProcessor.java:502)
	at com.softwarementors.extjs.djn.router.processor.standard.json.JsonRequestProcessor.processIndividualRequestsInThisThread(JsonRequestProcessor.java:150)
	at com.softwarementors.extjs.djn.router.processor.standard.json.JsonRequestProcessor.process(JsonRequestProcessor.java:133)
	at com.softwarementors.extjs.djn.router.RequestRouter.processJsonRequest(RequestRouter.java:83)
	at com.softwarementors.extjs.djn.servlet.DirectJNgineServlet.processRequest(DirectJNgineServlet.java:630)
	at com.softwarementors.extjs.djn.servlet.DirectJNgineServlet.doPost(DirectJNgineServlet.java:593)
	at org.sonatype.nexus.extdirect.internal.ExtDirectServlet.doPost(ExtDirectServlet.java:141)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:707)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
	at com.google.inject.servlet.ServletDefinition.doServiceImpl(ServletDefinition.java:286)
	at com.google.inject.servlet.ServletDefinition.doService(ServletDefinition.java:276)
	at com.google.inject.servlet.ServletDefinition.service(ServletDefinition.java:181)
	at com.google.inject.servlet.DynamicServletPipeline.service(DynamicServletPipeline.java:71)
	at com.google.inject.servlet.FilterChainInvocation.doFilter(FilterChainInvocation.java:85)
	at org.apache.shiro.web.servlet.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:112)
	at com.google.inject.servlet.FilterChainInvocation.doFilter(FilterChainInvocation.java:82)
	at org.apache.shiro.web.servlet.ProxiedFilterChain.doFilter(ProxiedFilterChain.java:61)
	at org.apache.shiro.web.servlet.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:118)
	at org.apache.shiro.web.servlet.ProxiedFilterChain.doFilter(ProxiedFilterChain.java:66)
	at org.apache.shiro.web.servlet.AdviceFilter.executeChain(AdviceFilter.java:108)
	at org.apache.shiro.web.servlet.AdviceFilter.doFilterInternal(AdviceFilter.java:137)
	at org.apache.shiro.web.servlet.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:125)
	at org.apache.shiro.web.servlet.ProxiedFilterChain.doFilter(ProxiedFilterChain.java:66)
	at org.apache.shiro.web.servlet.AdviceFilter.executeChain(AdviceFilter.java:108)
	at org.apache.shiro.web.servlet.AdviceFilter.doFilterInternal(AdviceFilter.java:137)
	at org.apache.shiro.web.servlet.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:125)
	at org.apache.shiro.web.servlet.ProxiedFilterChain.doFilter(ProxiedFilterChain.java:66)
	at org.apache.shiro.web.servlet.AbstractShiroFilter.executeChain(AbstractShiroFilter.java:449)
	at org.sonatype.nexus.security.SecurityFilter.executeChain(SecurityFilter.java:85)
	at org.apache.shiro.web.servlet.AbstractShiroFilter$1.call(AbstractShiroFilter.java:365)
	at org.apache.shiro.subject.support.SubjectCallable.doCall(SubjectCallable.java:90)
	at org.apache.shiro.subject.support.SubjectCallable.call(SubjectCallable.java:83)
	at org.apache.shiro.subject.support.DelegatingSubject.execute(DelegatingSubject.java:383)
	at org.apache.shiro.web.servlet.AbstractShiroFilter.doFilterInternal(AbstractShiroFilter.java:362)
	at org.sonatype.nexus.security.SecurityFilter.doFilterInternal(SecurityFilter.java:101)
	at org.apache.shiro.web.servlet.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:125)
	at com.google.inject.servlet.FilterChainInvocation.doFilter(FilterChainInvocation.java:82)
	at com.sonatype.nexus.licensing.internal.LicensingRedirectFilter.doFilter(LicensingRedirectFilter.java:108)
	at com.google.inject.servlet.FilterChainInvocation.doFilter(FilterChainInvocation.java:82)
	at com.codahale.metrics.servlet.AbstractInstrumentedFilter.doFilter(AbstractInstrumentedFilter.java:97)
	at com.google.inject.servlet.FilterChainInvocation.doFilter(FilterChainInvocation.java:82)
	at org.sonatype.nexus.internal.web.ErrorPageFilter.doFilter(ErrorPageFilter.java:68)
	at com.google.inject.servlet.FilterChainInvocation.doFilter(FilterChainInvocation.java:82)
	at org.sonatype.nexus.internal.web.EnvironmentFilter.doFilter(EnvironmentFilter.java:101)
	at com.google.inject.servlet.FilterChainInvocation.doFilter(FilterChainInvocation.java:82)
	at org.sonatype.nexus.internal.web.HeaderPatternFilter.doFilter(HeaderPatternFilter.java:98)
	at com.google.inject.servlet.FilterChainInvocation.doFilter(FilterChainInvocation.java:82)
	at com.google.inject.servlet.DynamicFilterPipeline.dispatch(DynamicFilterPipeline.java:104)
	at com.google.inject.servlet.GuiceFilter.doFilter(GuiceFilter.java:135)
	at org.sonatype.nexus.bootstrap.osgi.DelegatingFilter.doFilter(DelegatingFilter.java:73)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1634)
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:533)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:146)
	at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:548)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:132)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:257)
	at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:1595)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:255)
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1317)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:203)
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:473)
	at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:1564)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:201)
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1219)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:144)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:132)
	at com.codahale.metrics.jetty9.InstrumentedHandler.handle(InstrumentedHandler.java:175)
	at org.eclipse.jetty.server.handler.HandlerCollection.handle(HandlerCollection.java:126)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:132)
	at org.eclipse.jetty.server.Server.handle(Server.java:531)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:352)
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:260)
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:281)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:102)
	at org.eclipse.jetty.io.ChannelEndPoint$2.run(ChannelEndPoint.java:118)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.runTask(EatWhatYouKill.java:333)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:310)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:168)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:126)
	at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:366)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:762)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$2.run(QueuedThreadPool.java:680)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.IllegalArgumentException: A project ID is required for this service but could not be determined from the builder or the environment.  Please set a project ID using the builder.
	at com.google.common.base.Preconditions.checkArgument(Preconditions.java:135)
	at com.google.cloud.ServiceOptions.<init>(ServiceOptions.java:266)
	at com.google.cloud.datastore.DatastoreOptions.<init>(DatastoreOptions.java:102)
	at com.google.cloud.datastore.DatastoreOptions.<init>(DatastoreOptions.java:35)
	at com.google.cloud.datastore.DatastoreOptions$Builder.build(DatastoreOptions.java:89)
	at com.google.cloud.datastore.DatastoreOptions.getDefaultInstance(DatastoreOptions.java:158)
	at org.sonatype.nexus.blobstore.gcloud.internal.GoogleCloudDatastoreFactory.create(GoogleCloudDatastoreFactory.java:27)
	at org.sonatype.nexus.blobstore.gcloud.internal.DeletedBlobIndex.<init>(DeletedBlobIndex.java:43)
	at org.sonatype.nexus.blobstore.gcloud.internal.GoogleCloudBlobStore.init(GoogleCloudBlobStore.java:351)
	... 105 common frames omitted
```

Taking a brief look at things, I think there's a change between nexus 3.11 and 3.12 (and thus 3.13) that might be breaking this.  Because if I update the nexus dependency in the pom.xml to match the version I'm using, the project fails to compile with

```
[ERROR] COMPILATION ERROR :
[INFO] -------------------------------------------------------------
[ERROR] /Users/sherd/IdeaProjects/nexus-blobstore-google-cloud/src/main/java/org/sonatype/nexus/blobstore/gcloud/internal/GoogleCloudBlobStore.java:[74,14] 1. ERROR in /Users/sherd/IdeaProjects/nexus-blobstore-google-cloud/src/main/java/org/sonatype/nexus/blobstore/gcloud/internal/GoogleCloudBlobStore.java (at line 74)
	public class GoogleCloudBlobStore
	             ^^^^^^^^^^^^^^^^^^^^
The type GoogleCloudBlobStore must implement the inherited abstract method BlobStore.undelete(BlobStoreUsageChecker, BlobId, BlobAttributes, boolean)

[ERROR] Found 1 error and 0 warnings.
[INFO] 2 errors
[INFO] -------------------------------------------------------------
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 16.567 s
[INFO] Finished at: 2018-08-15T13:33:51-07:00
[INFO] Final Memory: 50M/405M
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.3:compile (default-compile) on project nexus-blobstore-google-cloud: Compilation failure: Compilation failure:
[ERROR] /Users/sherd/IdeaProjects/nexus-blobstore-google-cloud/src/main/java/org/sonatype/nexus/blobstore/gcloud/internal/GoogleCloudBlobStore.java:[74,14] 1. ERROR in /Users/sherd/IdeaProjects/nexus-blobstore-google-cloud/src/main/java/org/sonatype/nexus/blobstore/gcloud/internal/GoogleCloudBlobStore.java (at line 74)
[ERROR] 	public class GoogleCloudBlobStore
[ERROR] 	             ^^^^^^^^^^^^^^^^^^^^
[ERROR] The type GoogleCloudBlobStore must implement the inherited abstract method BlobStore.undelete(BlobStoreUsageChecker, BlobId, BlobAttributes, boolean)
[ERROR]
[ERROR] Found 1 error and 0 warnings.
[ERROR] -> [Help 1]
[ERROR]
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
```

--- a/README.md
+++ b/README.md
@@ -18,7 +18,7 @@ Nexus Repository Google Cloud Storage Blobstore
 [![Build Status](https://travis-ci.org/sonatype-nexus-community/nexus-blobstore-google-cloud.svg?branch=master)](https://travis-ci.org/sonatype-nexus-community/nexus-blobstore-google-cloud) [![Join the chat at https://gitter.im/sonatype/nexus-developers](https://badges.gitter.im/sonatype/nexus-developers.svg)](https://gitter.im/sonatype/nexus-developers?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge)
 
 This project adds [Google Cloud Object Storage](https://cloud.google.com/storage/) backed blobstores to Sonatype Nexus 
-Repository 3.  It allows Nexus Repository to store the components and assets in Google Cloud instead of a
+Repository 3.13 and later.  It allows Nexus Repository to store the components and assets in Google Cloud instead of a
 local filesystem.
 
 Contribution Guidelines

--- a/pom.xml
+++ b/pom.xml
@@ -20,7 +20,7 @@
   <parent>
     <groupId>org.sonatype.nexus.plugins</groupId>
     <artifactId>nexus-plugins</artifactId>
-    <version>3.11.0-01</version>
+    <version>3.13.0-01</version>
   </parent>
 
   <artifactId>nexus-blobstore-google-cloud</artifactId>

--- a/src/main/java/org/sonatype/nexus/blobstore/gcloud/internal/GoogleCloudBlobAttributes.java
+++ b/src/main/java/org/sonatype/nexus/blobstore/gcloud/internal/GoogleCloudBlobAttributes.java
@@ -146,6 +146,10 @@ public class GoogleCloudBlobAttributes
       properties.put(DELETED_ATTRIBUTE, Boolean.toString(deleted));
       properties.put(DELETED_REASON_ATTRIBUTE, getDeletedReason());
     }
+    else {
+      properties.remove(DELETED_ATTRIBUTE);
+      properties.remove(DELETED_REASON_ATTRIBUTE);
+    }
     return properties;
   }
 }

--- a/src/main/java/org/sonatype/nexus/blobstore/gcloud/internal/GoogleCloudBlobStore.java
+++ b/src/main/java/org/sonatype/nexus/blobstore/gcloud/internal/GoogleCloudBlobStore.java
@@ -461,6 +461,15 @@ public class GoogleCloudBlobStore
     return getBlobAttributes(blobId) != null;
   }
 
+  @Override
+  public boolean undelete(@Nullable final BlobStoreUsageChecker blobStoreUsageChecker,
+                          final BlobId blobId,
+                          final BlobAttributes blobAttributes,
+                          final boolean b)
+  {
+    return false;
+  }
+
   Blob createInternal(final Map<String, String> headers, BlobIngester ingester) {
     checkNotNull(headers);
 

--- a/src/main/java/org/sonatype/nexus/blobstore/gcloud/internal/GoogleCloudBlobStore.java
+++ b/src/main/java/org/sonatype/nexus/blobstore/gcloud/internal/GoogleCloudBlobStore.java
@@ -17,6 +17,7 @@ import java.io.InputStream;
 import java.nio.channels.Channels;
 import java.nio.file.Path;
 import java.util.Map;
+import java.util.Optional;
 import java.util.concurrent.atomic.AtomicInteger;
 import java.util.concurrent.locks.Lock;
 import java.util.stream.Stream;
@@ -38,6 +39,7 @@ import org.sonatype.nexus.blobstore.api.BlobStoreConfiguration;
 import org.sonatype.nexus.blobstore.api.BlobStoreException;
 import org.sonatype.nexus.blobstore.api.BlobStoreMetrics;
 import org.sonatype.nexus.blobstore.api.BlobStoreUsageChecker;
+import org.sonatype.nexus.common.log.DryRunPrefix;
 import org.sonatype.nexus.common.stateguard.Guarded;
 import org.sonatype.nexus.common.stateguard.StateGuardLifecycleSupport;
 import org.sonatype.nexus.logging.task.ProgressLogIntervalHelper;
@@ -63,6 +65,7 @@ import static com.google.common.base.Preconditions.checkState;
 import static com.google.common.cache.CacheLoader.from;
 import static java.lang.String.format;
 import static org.sonatype.nexus.blobstore.DirectPathLocationStrategy.DIRECT_PATH_ROOT;
+import static org.sonatype.nexus.blobstore.api.BlobAttributesConstants.HEADER_PREFIX;
 import static org.sonatype.nexus.common.stateguard.StateGuardLifecycleSupport.State.FAILED;
 import static org.sonatype.nexus.common.stateguard.StateGuardLifecycleSupport.State.NEW;
 import static org.sonatype.nexus.common.stateguard.StateGuardLifecycleSupport.State.STARTED;
@@ -118,15 +121,19 @@ public class GoogleCloudBlobStore
 
   private LoadingCache<BlobId, GoogleCloudStorageBlob> liveBlobs;
 
+  private final DryRunPrefix dryRunPrefix;
+
   @Inject
   public GoogleCloudBlobStore(final GoogleCloudStorageFactory storageFactory,
                               final BlobIdLocationResolver blobIdLocationResolver,
                               final GoogleCloudBlobStoreMetricsStore metricsStore,
-                              final GoogleCloudDatastoreFactory datastoreFactory) {
+                              final GoogleCloudDatastoreFactory datastoreFactory,
+                              final DryRunPrefix dryRunPrefix) {
     this.storageFactory = checkNotNull(storageFactory);
     this.blobIdLocationResolver = checkNotNull(blobIdLocationResolver);
     this.metricsStore = metricsStore;
     this.datastoreFactory = datastoreFactory;
+    this.dryRunPrefix = dryRunPrefix;
   }
 
   @Override
@@ -421,7 +428,7 @@ public class GoogleCloudBlobStore
   }
 
   /**
-   * @return the {@link BlobAttributes} for the blod, or null
+   * @return the {@link BlobAttributes} for the blob, or null
    * @throws BlobStoreException if an {@link IOException} occurs
    */
   @Override
@@ -464,9 +471,38 @@ public class GoogleCloudBlobStore
   @Override
   public boolean undelete(@Nullable final BlobStoreUsageChecker blobStoreUsageChecker,
                           final BlobId blobId,
-                          final BlobAttributes blobAttributes,
-                          final boolean b)
+                          final BlobAttributes attributes,
+                          final boolean isDryRun)
   {
+    checkNotNull(attributes);
+    String logPrefix = isDryRun ? dryRunPrefix.get() : "";
+    Optional<String> blobName = Optional.of(attributes)
+        .map(BlobAttributes::getProperties)
+        .map(p -> p.getProperty(HEADER_PREFIX + BLOB_NAME_HEADER));
+    if (!blobName.isPresent()) {
+      log.error("Property not present: {}, for blob id: {}, at path: {}", HEADER_PREFIX + BLOB_NAME_HEADER,
+          blobId, attributePath(blobId));
+      return false;
+    }
+    if (attributes.isDeleted() && blobStoreUsageChecker != null &&
+        blobStoreUsageChecker.test(this, blobId, blobName.get())) {
+      String deletedReason = attributes.getDeletedReason();
+      if (!isDryRun) {
+        attributes.setDeleted(false);
+        attributes.setDeletedReason(null);
+        try {
+          attributes.store();
+        }
+        catch (IOException e) {
+          log.error("Error while un-deleting blob id: {}, deleted reason: {}, blob store: {}, blob name: {}",
+              blobId, deletedReason, blobStoreConfiguration.getName(), blobName.get(), e);
+        }
+      }
+      log.warn(
+          "{}Soft-deleted blob still in use, un-deleting blob id: {}, deleted reason: {}, blob store: {}, blob name: {}",
+          logPrefix, blobId, deletedReason, blobStoreConfiguration.getName(), blobName.get());
+      return true;
+    }
     return false;
   }
 

--- a/src/test/java/org/sonatype/nexus/blobstore/gcloud/internal/GoogleCloudBlobStoreIT.groovy
+++ b/src/test/java/org/sonatype/nexus/blobstore/gcloud/internal/GoogleCloudBlobStoreIT.groovy
@@ -19,8 +19,13 @@ import org.sonatype.nexus.blobstore.BlobIdLocationResolver
 import org.sonatype.nexus.blobstore.DefaultBlobIdLocationResolver
 import org.sonatype.nexus.blobstore.PeriodicJobService
 import org.sonatype.nexus.blobstore.PeriodicJobService.PeriodicJob
+import org.sonatype.nexus.blobstore.api.Blob
+import org.sonatype.nexus.blobstore.api.BlobAttributes
 import org.sonatype.nexus.blobstore.api.BlobId
+import org.sonatype.nexus.blobstore.api.BlobStore
 import org.sonatype.nexus.blobstore.api.BlobStoreConfiguration
+import org.sonatype.nexus.blobstore.api.BlobStoreUsageChecker
+import org.sonatype.nexus.common.log.DryRunPrefix
 import org.sonatype.nexus.common.node.NodeAccess
 
 import com.google.cloud.storage.Blob.BlobSourceOption
@@ -64,6 +69,8 @@ class GoogleCloudBlobStoreIT
 
   GoogleCloudBlobStore blobStore
 
+  BlobStoreUsageChecker usageChecker = Mock()
+
   def setup() {
     config.attributes = [
         'google cloud storage': [
@@ -76,11 +83,14 @@ class GoogleCloudBlobStoreIT
 
     metricsStore = new GoogleCloudBlobStoreMetricsStore(periodicJobService, nodeAccess)
     // can't start metrics store until blobstore init is done (which creates the bucket)
-    blobStore = new GoogleCloudBlobStore(storageFactory, blobIdLocationResolver, metricsStore, datastoreFactory)
+    blobStore = new GoogleCloudBlobStore(storageFactory, blobIdLocationResolver, metricsStore, datastoreFactory,
+        new DryRunPrefix("TEST "))
     blobStore.init(config)
 
     blobStore.start()
     metricsStore.start()
+
+    usageChecker.test(_, _, _) >> true
   }
 
   def cleanup() {
@@ -126,6 +136,54 @@ class GoogleCloudBlobStoreIT
       results.size() == 2
       results.contains(new BlobId("${DIRECT_PATH_PREFIX}health-check/repo1/report.properties"))
       results.contains(new BlobId("${DIRECT_PATH_PREFIX}health-check/repo1/details/bootstrap.min.css"))
+  }
+
+  def "undelete successfully makes blob accessible"() {
+    given:
+      Blob blob = blobStore.create(new ByteArrayInputStream('hello'.getBytes()),
+          [ (BlobStore.BLOB_NAME_HEADER): 'foo',
+            (BlobStore.CREATED_BY_HEADER): 'someuser' ] )
+      assert blob != null
+      assert blobStore.delete(blob.id, 'testing')
+      BlobAttributes deletedAttributes = blobStore.getBlobAttributes(blob.id)
+      assert deletedAttributes.deleted
+      assert deletedAttributes.deletedReason == 'testing'
+
+    when:
+      !blobStore.undelete(usageChecker, blob.id, deletedAttributes, false)
+
+    then:
+      Blob after = blobStore.get(blob.id)
+      after != null
+      BlobAttributes attributesAfter = blobStore.getBlobAttributes(blob.id)
+      !attributesAfter.deleted
+  }
+
+  def "undelete does nothing when dry run is true"() {
+    given:
+      Blob blob = blobStore.create(new ByteArrayInputStream('hello'.getBytes()),
+          [ (BlobStore.BLOB_NAME_HEADER): 'foo',
+            (BlobStore.CREATED_BY_HEADER): 'someuser' ] )
+      assert blob != null
+      BlobAttributes attributes = blobStore.getBlobAttributes(blob.id)
+      assert blobStore.delete(blob.id, 'testing')
+      BlobAttributes deletedAttributes = blobStore.getBlobAttributes(blob.id)
+      assert deletedAttributes.deleted
+
+    when:
+      blobStore.undelete(usageChecker, blob.id, attributes, true)
+
+    then:
+      Blob after = blobStore.get(blob.id)
+      after == null
+      BlobAttributes attributesAfter = blobStore.getBlobAttributes(blob.id)
+      attributesAfter.deleted
+  }
+
+  def "undelete does nothing on non-existent blob"() {
+    expect:
+      BlobAttributes attributes = Mock()
+      !blobStore.undelete(usageChecker, new BlobId("nonexistent"), attributes, false)
   }
 
   def createFile(Storage storage, String path) {

--- a/src/test/java/org/sonatype/nexus/blobstore/gcloud/internal/GoogleCloudBlobStoreTest.groovy
+++ b/src/test/java/org/sonatype/nexus/blobstore/gcloud/internal/GoogleCloudBlobStoreTest.groovy
@@ -22,6 +22,7 @@ import org.sonatype.nexus.blobstore.api.BlobId
 import org.sonatype.nexus.blobstore.api.BlobStore
 import org.sonatype.nexus.blobstore.api.BlobStoreConfiguration
 import org.sonatype.nexus.blobstore.api.BlobStoreException
+import org.sonatype.nexus.common.log.DryRunPrefix
 
 import com.google.api.gax.paging.Page
 import com.google.cloud.datastore.Datastore
@@ -57,7 +58,7 @@ class GoogleCloudBlobStoreTest
       (BlobStore.CREATED_BY_HEADER): 'admin'
   ]
   GoogleCloudBlobStore blobStore = new GoogleCloudBlobStore(
-      storageFactory, blobIdLocationResolver, metricsStore, datastoreFactory)
+      storageFactory, blobIdLocationResolver, metricsStore, datastoreFactory, new DryRunPrefix("TEST "))
 
   def config = new BlobStoreConfiguration()
 

